File: RL_cilia_ball/value_iteration_cilia2ball.m
% value_iteration_cilia2ball.m
%
% Solve the small tabular 2-ball cilia problem by value iteration,
% then extract the optimal deterministic policy and analyze its cycles.
%
% Requires:
%   - setdefaultparams_ciliaball.m
%   - ciliaBallTabularEnv.m
%   - cilia_ball_reward.m
%
% Optional:
%   - position_from_angle.m (only for render/plots if you add them later)

clear; clc;
addpath('../RL_cilia_ball_tabular/');
% ============================================================
% USER SETTINGS
% ============================================================

gamma   = 0.99;      % discount factor
tol     = 1e-10;     % stopping tolerance for value iteration
maxIter = 10000;     % safety cap

% environment options
P = setdefaultparams_ciliaball;

envOpts = struct();
envOpts.reset_mode = 'fixed';
envOpts.precompute = true;

% choose one:
% envOpts.boundary = 'clip';
% envOpts.boundary = 'clip_penalty';
% envOpts.boundary = 'stay_penalty';

envOpts.boundary = 'clip_penalty';
envOpts.invalid_penalty = -0.1;

env = ciliaBallTabularEnv(P, envOpts);

fprintf('Built env with %d states and %d actions.\n', env.nStates, env.nActions);

if isempty(env.Rtable) || isempty(env.Ntable)
    error('Expected precomputed Rtable and Ntable. Set envOpts.precompute = true.');
end

R = env.Rtable;   % nStates x nActions
N = env.Ntable;   % nStates x nActions

nS = env.nStates;
nA = env.nActions;

% ============================================================
% VALUE ITERATION
% ============================================================

V = zeros(nS,1);
policy = ones(nS,1);

for it = 1:maxIter
    Vnew = zeros(nS,1);

    for s = 1:nS
        q_sa = zeros(nA,1);
        for a = 1:nA
            s2 = N(s,a);
            q_sa(a) = R(s,a) + gamma * V(s2);
        end

        [Vnew(s), policy(s)] = max(q_sa);
    end

    err = norm(Vnew - V, inf);

    if mod(it,50) == 0 || it == 1
        fprintf('Iter %4d | ||Vnew - V||_inf = %.3e\n', it, err);
    end

    V = Vnew;

    if err < tol
        fprintf('Value iteration converged at iter %d with err %.3e\n', it, err);
        break
    end
end

if it == maxIter
    warning('Reached maxIter before meeting tolerance.');
end

% ============================================================
% COMPUTE Q* FROM FINAL V
% ============================================================

Qstar = zeros(nS, nA);
for s = 1:nS
    for a = 1:nA
        s2 = N(s,a);
        Qstar(s,a) = R(s,a) + gamma * V(s2);
    end
end

% ============================================================
% FIND ALL UNIQUE CYCLES OF THE GREEDY POLICY
% ============================================================

cycles = find_all_cycles_from_policy(env, policy);
rankedCycles = rank_cycles_from_policy(env, policy, cycles);

fprintf('\nFound %d unique cycles under optimal greedy policy.\n', numel(rankedCycles));

disp('Top 10 cycles by average reward:');
for k = 1:min(10, numel(rankedCycles))
    C = rankedCycles{k};
    fprintf('\nCycle %d\n', k);
    fprintf('  avg reward   = %.8f\n', C.avg_reward);
    fprintf('  cycle length = %d\n', numel(C.states));
    fprintf('  states       = ');
    disp(C.states(:).');
end

% ============================================================
% SAVE RESULTS
% ============================================================

saveName = sprintf('valueIter_cilia2ball_%s_pen%0.2f_g%0.3f.mat', ...
    envOpts.boundary, envOpts.invalid_penalty, gamma);

save(saveName, ...
    'V', 'Qstar', 'policy', 'rankedCycles', ...
    'gamma', 'tol', 'maxIter', 'envOpts', 'P');

fprintf('\nSaved %s\n', saveName);

% ============================================================
% OPTIONAL: show best cycle in angle space
% ============================================================

if ~isempty(rankedCycles)
    bestCycle = rankedCycles{1}.states(:).';
    bestCycleClosed = [bestCycle, bestCycle(1)];

    cycle_phis = zeros(numel(bestCycleClosed), numel(P.Nstates));
    for j = 1:numel(bestCycleClosed)
        sub = env.state2sub(bestCycleClosed(j));
        phi = env.sub2phi(sub);
        cycle_phis(j,:) = phi(:).';
    end

    if size(cycle_phis,2) == 2
        figure;
        plot(cycle_phis(:,1), cycle_phis(:,2), 'o-', 'LineWidth', 2);
        xlabel('\phi_1');
        ylabel('\phi_2');
        title(sprintf('Best value-iteration cycle | len=%d | avg=%.6g', ...
            numel(bestCycle), rankedCycles{1}.avg_reward));
        grid on;
    end
end


% ============================================================
% LOCAL FUNCTIONS
% ============================================================

function cycles = find_all_cycles_from_policy(env, policy)
% Find all unique cycles under a deterministic policy on tabular states.

    nS = env.nStates;
    seenCycleKeys = containers.Map('KeyType','char','ValueType','int32');
    cycles = {};

    for s0 = 1:nS
        cyc = find_cycle_from_start(env, policy, s0);
        key = canonical_cycle_key(cyc);

        if ~isKey(seenCycleKeys, key)
            seenCycleKeys(key) = 1;
            cycles{end+1} = cyc(:).'; %#ok<SAGROW>
        end
    end
end

function cyc = find_cycle_from_start(env, policy, s0)
% Follow deterministic policy until a state repeats.

    visitedStep = zeros(env.nStates,1);
    traj = [];

    s = s0;
    t = 1;

    while visitedStep(s) == 0
        visitedStep(s) = t;
        traj(end+1) = s; %#ok<SAGROW>

        a = policy(s);
        s = env.Ntable(s,a);
        t = t + 1;
    end

    cycStart = visitedStep(s);
    cyc = traj(cycStart:end);
end

function key = canonical_cycle_key(cyc)
% Canonical representation of a cycle up to cyclic rotation.

    cyc = cyc(:).';
    L = numel(cyc);

    best = [];

    for k = 1:L
        rot = cyc([k:L, 1:k-1]);

        if isempty(best) || lex_less(rot, best)
            best = rot;
        end
    end

    key = sprintf('%d_', best);
end

function tf = lex_less(a, b)
% True if row vector a is lexicographically smaller than b.

    idx = find(a ~= b, 1, 'first');
    if isempty(idx)
        tf = false;
    else
        tf = a(idx) < b(idx);
    end
end

function ranked = rank_cycles_from_policy(env, policy, cycles)
% Rank cycles by average reward under the deterministic policy.

    ranked = cell(size(cycles));

    for i = 1:numel(cycles)
        cyc = cycles{i};
        L = numel(cyc);

        rewards = zeros(L,1);
        actions = zeros(L,1);

        for j = 1:L
            s = cyc(j);
            a = policy(s);
            actions(j) = a;
            rewards(j) = env.Rtable(s,a);
        end

        C.states = cyc(:).';
        C.actions = actions(:).';
        C.rewards = rewards(:).';
        C.avg_reward = mean(rewards);

        ranked{i} = C;
    end

    avgVals = cellfun(@(C) C.avg_reward, ranked);
    [~, perm] = sort(avgVals, 'descend');
    ranked = ranked(perm);
end

File: RL_cilia_python/old_top_level_copies/value_iteration_cilia_2_ball.py
import numpy as np

from cilia_2_ball_env import Cilia2BallEnv


def deterministic_next_state(env, state, action):
    """
    Deterministic next state, using precomputed tables if available.
    """
    i, j = int(state[0]), int(state[1])

    if env.precompute and env.next_state_table is not None:
        return tuple(env.next_state_table[i, j, action, :].tolist())

    trans = env.transition_info(np.array(state, dtype=int), action)
    return tuple(trans["next_state"].tolist())


def immediate_reward(env, state, action):
    """
    One-step reward, using precomputed tables if available.
    """
    i, j = int(state[0]), int(state[1])

    if env.precompute and env.flux_table is not None:
        return float(env.flux_table[i, j, action])

    trans = env.transition_info(np.array(state, dtype=int), action)
    return env.immediate_reward_from_transition(np.array(state, dtype=int), action, trans)


def value_iteration(env, gamma=0.99, tol=1e-10, max_iter=10000, verbose=True):
    """
    Value iteration for the deterministic finite MDP defined by env.

    Returns:
        V       : shape (n0, n1)
        Q       : shape (n0, n1, nA)
        policy  : shape (n0, n1)
        history : Bellman residual history
    """
    n0, n1 = int(env.n_bins[0]), int(env.n_bins[1])
    nA = env.action_space.n

    V = np.zeros((n0, n1), dtype=float)
    history = []

    for it in range(1, max_iter + 1):
        Vnew = np.zeros_like(V)

        for i in range(n0):
            for j in range(n1):
                qvals = np.empty(nA, dtype=float)

                for a in range(nA):
                    s2 = deterministic_next_state(env, (i, j), a)
                    r = immediate_reward(env, (i, j), a)
                    qvals[a] = r + gamma * V[s2]

                Vnew[i, j] = np.max(qvals)

        err = np.max(np.abs(Vnew - V))
        history.append(err)
        V = Vnew

        if verbose and (it == 1 or it % 50 == 0):
            print(f"Iter {it:4d} | ||Vnew - V||_inf = {err:.3e}")

        if err < tol:
            if verbose:
                print(f"Value iteration converged at iter {it} with err {err:.3e}")
            break
    else:
        print("Warning: reached max_iter before convergence.")

    Q = np.zeros((n0, n1, nA), dtype=float)
    policy = np.zeros((n0, n1), dtype=int)

    for i in range(n0):
        for j in range(n1):
            for a in range(nA):
                s2 = deterministic_next_state(env, (i, j), a)
                r = immediate_reward(env, (i, j), a)
                Q[i, j, a] = r + gamma * V[s2]

            policy[i, j] = int(np.argmax(Q[i, j, :]))

    return V, Q, policy, history


def canonical_cycle(cycle):
    """
    Canonical representation of a cycle, invariant under cyclic rotation.
    """
    cycle = [tuple(map(int, s)) for s in cycle]
    n = len(cycle)
    rots = [tuple(cycle[k:] + cycle[:k]) for k in range(n)]
    return min(rots)


def find_cycle(env, policy, start_state):
    """
    Deterministic cycle under a fixed policy.
    """
    visited = {}
    trajectory = []

    state = tuple(map(int, start_state))
    t = 0

    while state not in visited:
        visited[state] = t
        trajectory.append(state)

        i, j = state
        a = int(policy[i, j])
        state = deterministic_next_state(env, state, a)
        t += 1

    cycle_start = visited[state]
    return trajectory[cycle_start:]


def find_all_cycles(env, policy):
    """
    Find all unique recurrent cycles under the deterministic policy.
    """
    n0, n1 = int(env.n_bins[0]), int(env.n_bins[1])

    seen = set()
    unique_cycles = []

    for i in range(n0):
        for j in range(n1):
            cyc = find_cycle(env, policy, (i, j))
            canon = canonical_cycle(cyc)
            if canon not in seen:
                seen.add(canon)
                unique_cycles.append(list(canon))

    return unique_cycles


def rank_cycles(env, policy, cycles):
    """
    Rank cycles by average one-step reward under the policy.
    """
    results = []

    for cycle in cycles:
        rewards = []
        actions = []

        for state in cycle:
            i, j = state
            a = int(policy[i, j])
            r = immediate_reward(env, state, a)
            rewards.append(r)
            actions.append(a)

        avg_reward = float(np.mean(rewards))
        results.append(
            {
                "cycle": cycle,
                "actions": actions,
                "rewards": rewards,
                "avg_reward": avg_reward,
                "length": len(cycle),
            }
        )

    results.sort(key=lambda d: (-d["avg_reward"], -d["length"]))
    return results


def print_top_cycles(cycles_ranked, top_k=10):
    for k, C in enumerate(cycles_ranked[:top_k], start=1):
        print(f"\nCycle {k}:")
        print(f"  Average Reward: {C['avg_reward']:.6f}")
        print(f"  Cycle Length:   {C['length']}")
        print(f"  Cycle States:   {C['cycle']}")
        print(f"  Actions:        {C['actions']}")


if __name__ == "__main__":
    # ------------------------------------------------------------
    # Change this to compare the two boundary models
    # ------------------------------------------------------------
    boundary_mode = "clip_penalty"   # or "stay_penalty"

    env = Cilia2BallEnv(
        max_steps=500,
        precompute=True,
        boundary_mode=boundary_mode,
        invalid_penalty=-0.1,
        reward_rescale=100.0,
        n_bins=[11, 21],
        angle_mins=[-np.pi / 4, -np.pi / 2],
        angle_maxs=[ np.pi / 4,  np.pi / 2],
    )

    gamma = 0.99

    V, Q, policy, history = value_iteration(
        env,
        gamma=gamma,
        tol=1e-10,
        max_iter=10000,
        verbose=True,
    )

    raw_cycles = find_all_cycles(env, policy)
    cycles = rank_cycles(env, policy, raw_cycles)

    print("\n===================================")
    print(f"Boundary mode: {boundary_mode}")
    print(f"Found {len(cycles)} unique cycles")
    print("Top 10 cycles by average reward:")
    print_top_cycles(cycles, top_k=10)

    save_name = (
        f"value_iteration_cilia_2_ball_"
        f"{boundary_mode}_"
        f"bins{env.n_bins[0]}x{env.n_bins[1]}_"
        f"g{gamma:.3f}.npy"
    )

    out = {
        "V": V,
        "Q": Q,
        "policy": policy,
        "history": np.array(history),
        "cycles": np.array(cycles, dtype=object),
        "gamma": gamma,
        "boundary_mode": boundary_mode,
        "invalid_penalty": env.invalid_penalty,
        "reward_rescale": env.reward_rescale,
        "n_bins": np.array(env.n_bins, dtype=int),
        "angle_mins": np.array(env.angle_mins, dtype=float),
        "angle_maxs": np.array(env.angle_maxs, dtype=float),
        "dangle": np.array(env.dangle, dtype=float),
    }

    np.save(save_name, out, allow_pickle=True)
    print(f"\nSaved {save_name}")